### DISCUSSION TITLE
Plans for supporting the libxdp protocol for attaching XDP component programs?

### DISCUSSION THREAD

**Discussion by thediveo** [cilium/ebpf#940] [Q&A] [answered]
Please forgive me in case I've missed this feature, but looking especially at `link.go` I'm under the impression that this isn't supported yet.

[libxdp's README.org](https://github.com/xdp-project/xdp-tools/blob/master/lib/libxdp/README.org) and especially [libxdp/protocol.org](https://github.com/xdp-project/xdp-tools/blob/master/lib/libxdp/protocol.org) describe a protocol that allows to modularize XDP programs into what I think is called "components". As this is done in user space, libxdp documents the protocol behind it, not least that other implementation (such as a Go native, non-libxdp/cgo) can successfully interoperate with libxdp.

Are there any plans on supporting the component feature in the cilium/ebpf module?

**Comment by ti-mo** [comment, marked as answer]
libxdp is a project built on top of libbpf, so another Go library could implement the same approach on top of ebpf-go. libxdp takes a highly opinionated approach, so for the same reason libxdp exists as a separate library from libbpf, ebpf-go is unlikely to ever implement a similar concept directly.

@dylandreimerink was working on an AF_XDP implementation some time ago, not sure if this is related.

**Comment by wenlxie** [comment]
@ti-mo  @dylandreimerink Does libxdp (for attach multiple xdp prog) supported already?  Thanks 
With this, we can support to integrate 3rd party xdp programs like xdpcap to cilium xdp.

**Reply by dylandreimerink** [reply]
We do not support it "out of the box". However, you should be able to use the library to perform all of the steps. If I recall correctly you have to load your XDP program as an [extention type](https://github.com/cilium/ebpf/blob/3a926eb8a6eeced1aa8a79dbdd3851ea1c93cf2c/types.go#L150) program making sure to specify the correct kernel data type to attach to and [attach program](https://github.com/cilium/ebpf/blob/3a926eb8a6eeced1aa8a79dbdd3851ea1c93cf2c/prog.go#L105). And finally do the actual replacing of a dummy function in the dispatcher with your XDP program with a [freplace link](https://github.com/cilium/ebpf/blob/3a926eb8a6eeced1aa8a79dbdd3851ea1c93cf2c/link/tracing.go#L30).

I suggest taking a look at the libxdp code to see how everything needs to be wired up, but all of the building blocks are there I believe.

**Reply by wenlxie** [reply]
@dylandreimerink thanks, I will take a look

**Reply by wenlxie** [reply]
@dylandreimerink  
I am trying to have a Poc for  cilium to support multiple xdp progs which works like xdp-tools
https://github.com/xdp-project/xdp-tools/tree/master/xdp-loader

I loaded the main prog xdp-dispatcher as XDP type. 
```

	dispatcher, err := bpf.LoadCollectionSpec(option.Config.StateDir + "/xdp-dispatcher.o")
	if err != nil {
		return nil, fmt.Errorf("loading eBPF xdp-dispatcher ELF: %w", err)
	}
	l.Debugf("Loaded spec for xdp dispatcher:%+v", dispatcher)

	maincoll, err := bpf.LoadCollection(dispatcher, ebpf.CollectionOptions{})
	if err != nil {
		l.WithError(err).Errorf("Couldn't load xdp-dispatcher, err:%v", err)
		return nil, fmt.Errorf("Couldn't load xdp-dispatcher, err:%v", err)
	}

```

Then I load bpf_xdp.o of cilium as extention type.
1. LoadSpec from the objPath
2. Set the cil_xdp_entry prog to be extention type:
```
			prog.AttachTarget = maincoll.Programs["xdp_dispatcher"]
			prog.Type = ebpf.Extension
			prog.AttachType = ebpf.AttachNone
			prog.AttachTo = "prog0"
			spec.Programs[name] = prog
```
3. 
But when I load the bpf_xdp.o to kernel by bpf.LoadCollection, it shows the error 
```
level=debug msg="Loading Collection into kernel" device=eth0 ifindex=86 objPath=/var/run/cilium/state/bpf_xdp.o subsys=datapath-loader
Verifier error: program cil_xdp_entry: load program: invalid argument
Verifier log: load program: invalid argument:
	processed 271 insns (limit 1000000) max_states_per_insn 0 total_states 22 peak_states 22 mark_read 6
level=fatal msg="Failed to compile XDP program" error="loading eBPF collection into the kernel: program cil_xdp_entry: load program: invalid argument" subsys=datapath-loader
```

4.  Then  I try to load a simple xdp prog as the extention type, then there is no invalid argument errors. 

5.  So  I am not sure what's the error in bpf_xdp.o, any suggestions?

6.  Another question is:
bpf_xdp.o has lots of tailcalls, do you know how to set  the   type/attach type/AttachTarget of these tailed programs? 

Appreciate that if you can help to have a check. Thanks
